kore/src/Kore/Builtin/IO.hs

```diff
@@ -74,29 +113,409 @@ symbolVerifiers =
 builtinFunctions :: Text -> Maybe BuiltinAndAxiomSimplifier
 builtinFunctions key
     | key == logStringKey = Just $ Builtin.functionEvaluator evalLogString
+    | key == openKey = Just $ Builtin.functionEvaluator evalOpen
+    | key == getcKey = Just $ Builtin.functionEvaluator evalGetc
+    | key == readKey = Just $ Builtin.functionEvaluator evalRead
+    | key == closeKey = Just $ Builtin.functionEvaluator evalClose
+    | key == writeKey = Just $ Builtin.functionEvaluator evalWrite
     | otherwise = Nothing
 
+mkDotk ::
+    MonadSimplify m =>
+    TermLike.InternalVariable variable =>
+    TermLike.Sort ->
+    m (Pattern.Pattern variable)
+mkDotk sort = do
+    tools <- askMetadataTools
+    let ident = implicitId "dotk"
+    return $
+        Pattern.fromTermLike $
+            TermLike.mkApplySymbol
+                Symbol
+                    { symbolConstructor = ident
+                    , symbolParams = []
+                    , symbolSorts =
+                        ApplicationSorts
+                            { applicationSortsOperands = []
+                            , applicationSortsResult = sort
+                            }
+                    , symbolAttributes = Tools.symbolAttributes tools ident
+                    }
+                []
+
 evalLogString :: Builtin.Function
 evalLogString _ _ [] = Builtin.wrongArity logStringKey
 evalLogString _ sort [InternalString_ InternalString{internalStringValue}] = do
-    tools <- askMetadataTools
     infoUserLog internalStringValue
-    return $ Pattern.fromTermLike $ TermLike.mkApplySymbol (dotk tools) []
-  where
-    dotk tools =
-        Symbol
-            { symbolConstructor = ident
-            , symbolParams = []
-            , symbolSorts =
-                ApplicationSorts
-                    { applicationSortsOperands = []
-                    , applicationSortsResult = sort
-                    }
-            , symbolAttributes = Tools.symbolAttributes tools ident
-            }
-    ident = implicitId "dotk"
+    mkDotk sort
 evalLogString _ _ [_] = empty
 evalLogString _ _ (_ : _ : _) = Builtin.wrongArity logStringKey
 
 logStringKey :: IsString s => s
 logStringKey = "IO.logString"
+
+returnErrnoIfMinus1OrRetry :: (Eq a, Num a) => IO a -> IO (Either Errno a)
+returnErrnoIfMinus1OrRetry f =
+    do
+        res <- f
+        if res == -1
+            then do
+                err <- getErrno
+                if err == eINTR
+                    then returnErrnoIfMinus1OrRetry f
+                    else return $ Left err
+            else return $ Right res
+
+parseMode :: Text -> Maybe (OpenMode, OpenFileFlags, Maybe FileMode)
+parseMode = \case
+    "r" -> Just (ReadOnly, defaultFileFlags, Nothing)
+    "r+" -> Just (ReadWrite, defaultFileFlags, Nothing)
+    "w" -> Just (WriteOnly, defaultFileFlags{trunc = True}, Just 0o666)
```

Comment:
`0o666` is a magical constant I found here https://github.com/ghc/ghc/blob/311251543f2e37af4a121e58028bfc46267a7fc9/libraries/base/GHC/IO/FD.hs#L238 :P

Reply:
Can you add this as a comment in the code? Will probably be useful later